testing/file_test/file_test_base.cpp

```diff
@@ -514,6 +517,82 @@ struct SplitState {
   int file_index = 0;
 };
 
+// Replaces the keyword at the given position. Returns the position to start a
+// find for the next keyword.
+static auto ReplaceContentKeywordAt(std::string* content, size_t keyword_pos,
+                                    llvm::StringRef test_name, int* lsp_id)
+    -> ErrorOr<size_t> {
+  auto keyword = llvm::StringRef(*content).substr(keyword_pos);
+
+  // Line replacements aren't handled here.
+  static constexpr llvm::StringLiteral Line = "[[@LINE";
+  if (keyword.starts_with(Line)) {
+    // Just move past the prefix to find the next one.
+    return keyword_pos + Line.size();
+  }
+
+  // Replaced with the actual test name.
+  static constexpr llvm::StringLiteral TestName = "[[@TEST_NAME]]";
+  if (keyword.starts_with(TestName)) {
+    content->replace(keyword_pos, TestName.size(), test_name);
+    return keyword_pos + test_name.size();
+  }
+
+  // Reformatted as an LSP call with headers.
+  static constexpr llvm::StringLiteral Lsp = "[[@LSP:";
+  if (keyword.starts_with(Lsp)) {
+    auto method_start = keyword_pos + Lsp.size();
+    auto method_end = content->find(":", method_start);
+    if (method_end == std::string::npos) {
+      return ErrorBuilder()
+             << "Missing `:` after method name for `" << Lsp << "`";
+    }
+    auto method = content->substr(method_start, method_end - method_start);
+
+    static constexpr llvm::StringLiteral LspEnd = "]]";
+    auto keyword_end = content->find("]]", method_end + 1);
```

Comment:
It'd be nice to do this search from `method_start` instead of from `method_end + 1` and reject if the `]]` is before the `:`, so that errors like this are rejected:
```
[[@LSP:exit]]
[[@LSP:something-else:blah]]
```
... rather than turning into a method of `exit]]\n[[@LSP` and extra content of `something-else:blah`.

Alternatively, if the `]]` is before the next `:`, we could treat the extra content as being empty?